pkg/controller/scheduledscaling/scheduled_scaling.go

```diff
@@ -183,30 +191,220 @@ func (c *Controller) runOnce(ctx context.Context) error {
 	return nil
 }
 
-func (c *Controller) scheduleActive(spec v1.ScalingScheduleSpec) (bool, error) {
+func (c *Controller) runOnce(ctx context.Context) error {
+	schedulesInterface := c.scalingScheduleStore.List()
+	schedules := make([]*v1.ScalingSchedule, 0, len(schedulesInterface))
+	for _, scheduleInterface := range schedulesInterface {
+		schedule, ok := scheduleInterface.(*v1.ScalingSchedule)
+		if !ok {
+			return ErrNotScalingScheduleFound
+		}
+		schedules = append(schedules, schedule)
+	}
+
+	clusterschedulesInterface := c.clusterScalingScheduleStore.List()
+	clusterschedules := make([]*v1.ClusterScalingSchedule, 0, len(clusterschedulesInterface))
+	for _, scheduleInterface := range clusterschedulesInterface {
+		schedule, ok := scheduleInterface.(*v1.ClusterScalingSchedule)
+		if !ok {
+			return ErrNotScalingScheduleFound
+		}
+		clusterschedules = append(clusterschedules, schedule)
+	}
+
+	err := c.updateStatus(ctx, schedules, clusterschedules)
+	if err != nil {
+		return fmt.Errorf("failed to update status: %w", err)
+	}
+
+	log.Info("Adjusting scaling")
+	err = c.adjustScaling(ctx, schedules, clusterschedules)
+	if err != nil {
+		return fmt.Errorf("failed to adjust scaling: %w", err)
+	}
+
+	return nil
+}
+
+// activeScheduledScaling returns a map of the current active schedules and the
+// max value per schedule.
+func (c *Controller) activeScheduledScaling(schedules []*v1.ScalingSchedule, clusterschedules []*v1.ClusterScalingSchedule) map[string]int64 {
+	currentActiveSchedules := make(map[string]int64)
+
+	// ScalingSchedules
+	for _, schedule := range schedules {
+		activeSchedules, err := c.activeSchedules(schedule.Spec)
+		if err != nil {
+			log.Errorf("Failed to check for active schedules in ScalingSchedule %s/%s: %v", schedule.Namespace, schedule.Name, err)
+			continue
+		}
+
+		if len(activeSchedules) == 0 {
+			continue
+		}
+
+		maxValue := int64(0)
+		for _, activeSchedule := range activeSchedules {
+			if activeSchedule.Value > maxValue {
+				maxValue = activeSchedule.Value
+			}
+		}
+		currentActiveSchedules[schedule.Namespace+"/"+schedule.Name] = maxValue
+	}
```

Comment:
Since the function `c.activeSchedules` expects a `v1.Schedule` type that's the same for both ScalingSchedule and ClusterScalingSchedule, can't we avoid the code duplication here easily? 

Maybe, having a function that takes `[]v1.Schedule` type as input and return the max value of the active schedule in there? Kind of how we detect the `activeSchedules`.

Reply:
The problem is we have a list of `[]*v1.ScalingSchedule` and `[]*v1.ClusterScalingSchedule` which we need to unpack into `[]v1.Schedule`. Since go can't understand that different types with the same field name/type (in this case `.Spec`) could be handled the same we at least need to run through the `[]*v1.ScalingSchedule` and `[]*v1.ClusterScalingSchedule` once to unpack it.

I have tried to create an interface to handle scalingSchedule and ClusterScalingSchedule in a cosistent way. Please check if that is more readable now.

Reply:
Looks great now 🙂 Thanks!